### PR TITLE
Implement SSH proxy over WebSockets (based on master branch)

### PR DESCRIPTION
**Pull request by achimnol**
This allows access to servers inside remote private networks.

### CODE REVIEW

File: geofront/server.py [hunk 1]
```diff
@@ -1033,6 +1042,58 @@ def authorize_remote(token_id: str, alias: str):
     )
 
 
+@sockets.route('/ws/tokens/<token_id:token_id>/remotes/<alias>/ssh/')
+def proxy_ssh(ws, token_id: str, alias: str):
```
Comment (dahlia) [inline geofront/server.py]: This function need some docs.

File: geofront/server.py [hunk 2]
```diff
@@ -1063,7 +1124,8 @@ def main_parser() -> argparse.ArgumentParser:  # pragma: no cover
                         help='IP address of a client allowed to override '
                              'url_scheme via the X-Forwarded-Proto header. '
                              'useful when it runs behind reverse proxy. '
-                             '-d/--debug option disables this option')
+                             '-d/--debug option disables this option. '
+                             '(deprecated)')
```
Comment (dahlia) [inline geofront/server.py]: If this option has any behavior anymore it should be removed (and recorded to changelog) rather deprecated.

File: geofront/server.py [hunk 3]
```diff
@@ -1033,6 +1042,58 @@ def authorize_remote(token_id: str, alias: str):
...
+        response.status_code = 403
+        return response
+
+    ssh_sock = create_connection((remote.host, remote.port))
+
+    def pipe():
+        while True:
+            try:
+                reply = ssh_sock.recv(4096)
+            except OSError:
```
Comment (dahlia) [inline geofront/server.py]: Isn't there a more specific exception type for this?

File: setup.py [hunk 4]
```diff
@@ -18,13 +18,14 @@ def readme():
...
     'cryptography >= 1.4',
+    'gevent >= 1.1.2',
     # indirect dependency thorugh paramiko; just for version constraint (>=1.4)
     'paramiko >= 2.0.1',
     'Werkzeug >= 0.11',
     'oauthlib[rsa, signedtoken] >= 1.1.1, < 2.0.0',
     'Flask >= 0.10.1',
+    'Flask-Sockets >= 0.2.1',
     'apache-libcloud >= 1.1.0',
-    'waitress >= 1.0.2, < 2.0.0'
```
Comment (dahlia) [inline setup.py]: Prerequisites documented in README.rst also need to be updated.

Comment (achimnol) [inline setup.py]: Prerequisites in README of the *docker* branch does not mention individual dependencies but only Docker or a link to the separate document about barebone installation.

Comment (achimnol) [inline setup.py]: Ah, but this is the *ssh-websocket-proxy* branch! :scream:

File: geofront/server.py (hunk 3 above)
Comment (achimnol) [inline geofront/server.py]: It is to handle [this](https://github.com/gevent/gevent/blob/f3acb176d0f0f1ac797b50e44a5e03726f687c53/src/gevent/_fileobjectcommon.py#L11-L13) and there is no intrinsic Exception classes corresponding to EBADF.

File: geofront/server.py (hunk 1 above)
Comment (achimnol) [inline geofront/server.py]: Added, but not shown in the docs currently due to kennethreitz/flask-sockets#50.

File: docs/changes.rst [hunk 5]
```diff
@@ -1,6 +1,28 @@
...
+  via SSH tunneling over an HTTP WebSocket.  [:issue:`21`]
+
+  - Added ``addresser`` and ``filter`` options to :class:`CloudRemoteSet
+    <geofront.backends.cloud.CloudRemoteSet>` constructor, which allow
+    customization of server addressing and filtering schemes.  The default
+    is still to use public IPs, so you would wish to give lambda functions that
+    return private IPs from the :class:`libcloud.compute.base.Node` object
+    for use with SSH tunneling.
+  - Removed ``--trusted-proxy`` option from the :program:`geofront-server`
+    command.
```
Comment (dahlia) [inline docs/changes.rst]: These two changes seem to have no direct relationship to SSH tunneling.  Shouldn't these two bullet points be dedented?

File: geofront/backends/cloud.py [hunk 6]
```diff
@@ -83,6 +83,19 @@ class CloudRemoteSet(collections.abc.Mapping):
...
+                      is used by default.
+    :type addresser:
+        :class:`~typing.Callable`\ [[:class:`libcloud.compute.base.Node`],
+                                    :class:`str`]
+    :param filter: A function to decide if the given node should be included as
+                   a remote or not.  By default it checks if the node has any
+                   public IP.
+    :type filter:
+        :class:`~typing.Callable`\ [[:class:`libcloud.compute.base.Node`],
+                                    :class:`bool`]
```
Comment (dahlia) [inline geofront/backends/cloud.py]: Added two options should be also logged using [`.. versionchanged:: 0.5.0` directive][1] below.

[1]: http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionchanged

File: geofront/server.py [hunk 7]
```diff
@@ -959,6 +969,35 @@ def list_remotes(token_id: str):
...
+    """Retrieve the information of the given remote alias.
+
+    :param token_id: the token id that holds the identity
+    :type token_id: :class:`str`
+    :param alias: the alias of the remote to access
+    :type alias: :class:`str`
+    :status 200: when successfully granted a temporary authorization
+    :status 404: (``not-found``) when there's no such remote
+
+    """
```
Comment (dahlia) [inline geofront/server.py]: [`.. versionadded:: 0.5.0` directive][1] is missing.

[1]: http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded

File: geofront/server.py [hunk 8]
```diff
@@ -1033,6 +1073,75 @@ def authorize_remote(token_id: str, alias: str):
...
+    :param ws: The WebSocket connection object
+    :param token_id: the token id that holds the identity
+    :type token_id: :class:`str`
+    :param alias: the alias of the remote to access
+    :type alias: :class:`str`
+    :status 101: good to upgrade to a WebSocket connection.
+    :status 403: when the permission policy denies access
+    :status 404: (``not-found``) when there's no such remote
+
+    """
```
Comment (dahlia) [inline geofront/server.py]: [`.. versionadded:: 0.5.0` directive][1] is missing.

[1]: http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded

**Comment by achimnol** [comment]
Thanks!
I will work on the requested changes soon.

**Comment by achimnol** [comment]
Separated #30.

**Comment by achimnol** [comment]
Removed the commit corresponding to #30 and rebased against the upstream master again.

**Comment by achimnol** [comment]
Rebased again!

**Pull request merged**